### DISCUSSION TITLE
LDAP connection pooling for user authentication

### DISCUSSION THREAD

**Discussion by dteleguin** [keycloak/keycloak#21365] [Ideas]
In Keycloak, LDAP authentication [does not use](https://github.com/keycloak/keycloak/blob/main/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/store/ldap/LDAPOperationManager.java#L501) connection pooling, with the substantiation being:

> Never use connection pool to prevent password caching

Could we revisit this? In deployments with high volume of LDAP authentications, establishing a LDAP connection every time could have a significant performance impact (especially in the case of TLS connections).

Technically, it should be possible to reset the connection state to anonymous before returning it to the pool (and erase password from memory). However, the logic of that pool might differ from that of the pool we're using for LDAP data synchronization, and with JNDI we can have only one default pool (correct me if I'm wrong). This could be solved e.g. with [Apache LDAP API](https://directory.apache.org/api/) (which is also way more powerful and versatile than JNDI).

On a related note, have we ever considered migrating from JNDI to Apache LDAP API in the long term?

**Comment by alphanarslan** [comment]
Hello team. We are having same performance problem with high volume LDAP authentication. Do you have a suggestion to improve the performance?

**Comment by mcarbonneaux** [comment]
without pool it's generate bind/unbind for every ldap request ... can you use pool for every read operation and only no pool on authentication ?
it's possible tu use pool for synchronization ?
